src/nccl_ofi_freelist.c

```diff
@@ -162,20 +210,32 @@ int nccl_ofi_freelist_add(nccl_ofi_freelist_t *freelist,
 	   structure at the end of the allocation so that large
 	   buffers are more likely to be page aligned (or aligned to
 	   their size, as the case may be). */
-	buffer = malloc(sizeof(struct nccl_ofi_freelist_block_t) +
-			(freelist->entry_size * allocation_count));
-	if (!buffer) {
-		NCCL_OFI_WARN("freelist extension malloc failed: %s", strerror(errno));
-		return -ENOMEM;
+	block_mem_size = freelist_block_mem_size_full_pages(freelist->entry_size, allocation_count);
+	buffer = mmap(NULL, block_mem_size, PROT_READ | PROT_WRITE,
+		    MAP_PRIVATE | MAP_ANON, -1, 0);
+	if (OFI_UNLIKELY(buffer == MAP_FAILED)) {
+		NCCL_OFI_WARN("Unable to map MR buffer (%d %s)",
+			      errno, strerror(errno));
+		buffer = NULL;
+		return -errno;
 	}
 
 	block = (struct nccl_ofi_freelist_block_t*)(buffer + (freelist->entry_size * allocation_count));
 	block->memory = buffer;
+	block->memory_size = block_mem_size;
 	block->next = freelist->blocks;
 
 	if (freelist->regmr_fn) {
-		ret = freelist->regmr_fn(freelist->regmr_opaque, block->memory,
-					 freelist->entry_size * allocation_count,
+		/*
+		 * Note that we are registering the entire block
+		 * memory buffer rather than only its entries,
+		 * including the block metadata structure.  Because we
+		 * require all internally registered memory regions
+		 * buffers cover full system memory pages. For more
+		 * information, see reg_internal_mr_ep().
```

Comment:
`reg_internal_mr_ep()` function doesn't exist in the context of this commit. It is introduced only by the following commit in patch-series.